fncas/fncas_optimize.h

```diff
@@ -0,0 +1,266 @@
+  fncas::reset_internals_singleton();
+  const size_t dim = starting_point.size();
+  const fncas::x gradient_helper(dim);
+  fncas::f_intermediate fi(F::compute(gradient_helper));
+  fncas::g_intermediate gi = fncas::g_intermediate(gradient_helper, fi);
+  std::vector<double> current_point(starting_point);
+
+  for (size_t iteration = 0; iteration < max_steps_; ++iteration) {
+    const auto g = gi(current_point);
+    const auto try_step = [&dim, &current_point, &fi, &g](double step) -> std::pair<double, std::vector<double>> {
+      std::vector<double> candidate(dim);
+      for (size_t i = 0; i < dim; ++i) {
+        candidate[i] = current_point[i] - g.gradient[i] * step;
+      }
+      const double value = fi(candidate);
+      return std::make_pair(value, candidate);
+    };
+    current_point = std::min(try_step(0.01), std::min(try_step(0.05), try_step(0.2))).second;
+  }
+
+  return OptimizationResult{fi(current_point), current_point};
+}
+
+
+template<class F> 
+OptimizationResult GradientDescentOptimizerBT<F>::Optimize(const std::vector<double>& starting_point) {
+  fncas::reset_internals_singleton();
+  const size_t dim = starting_point.size();
+  const fncas::x gradient_helper(dim);
+  fncas::f_intermediate fi(F::compute(gradient_helper));
+  fncas::g_intermediate gi = fncas::g_intermediate(gradient_helper, fi);
+  std::vector<double> current_point(starting_point);
+
+  for (size_t iteration = 0; iteration < max_steps_; ++iteration) {
+    const auto current_f_gradf = gi(current_point);
+    const double current_f_value = current_f_gradf.value;
+    double squared_gradient = 0.0;
+    std::vector<double> test_point(dim);
+
+    for (size_t i = 0; i < dim; ++i) {
+      test_point[i] = current_point[i] - current_f_gradf.gradient[i];
+      squared_gradient += current_f_gradf.gradient[i] * current_f_gradf.gradient[i];
+    }
+
+    double t = 1.0;
+    double test_f_value = fi(test_point);
+    
+    // Naive backtracking with strict limit on the number of steps.
+    size_t bt_iteration = 0;
+    while (test_f_value > (current_f_value - bt_alpha_ * t * squared_gradient)) {
+      t *= bt_beta_;
+      for (size_t i = 0; i < dim; ++i) {
+        test_point[i] = current_point[i] - t * current_f_gradf.gradient[i];
+      }
+      test_f_value = fi(test_point);
+      if (bt_iteration++ == bt_max_steps_) {
+        break;
+      }
+    }
+    current_point = test_point;
+
+    // Simple early stopping by the norm of gradient.
+    if (std::sqrt(squared_gradient) < grad_eps_) {
+      break;
+    }
+  }
+
+ return OptimizationResult{fi(current_point), current_point};
+}
+
+template<class F>
+double ConjugateGradientOptimizer<F>::PolakRibiere(const std::vector<double> &g,
+                                                   const std::vector<double> &g_prev) {
+  double beta = (fncas::L2Norm(g) - fncas::DotProduct(g, g_prev)) / fncas::L2Norm(g_prev);
+  if (!std::isinf(beta) && !std::isnan(beta)) {
```

Comment:
I'd probably:
1. move this check into `mathutil.h` (Also, check out `(x != 0.0 && std::isnormal(x)`, or some other form of stricter logic, which might be a better fit here), and
2. add these infinity-related check to the BT algorithm as well.